File: 백준/Gold/14502. 연구소/연구소.py
# https://jie0025.tistory.com/209

from collections import deque
import copy

n, m = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(n)]

def dfs(level):
    if level==3:
        bfs()
        return
    for i in range(n):
        for j in range(m):
            if arr[i][j]==0:
                arr[i][j] = 1
                dfs(level+1)
                arr[i][j] = 0

def bfs():
    q = deque()
    arr2 = copy.deepcopy(arr)
    for i in range(n):
        for j in range(m):
            if arr2[i][j]==2:
                q.append((i, j))

    while q:
        y, x = q.popleft()

        for dy, dx in ((0, -1), (0, 1), (1, 0), (-1, 0)):
            ny, nx = y+dy, x+dx
            if 0<=ny<n and 0<=nx<m:
                if arr2[ny][nx]==0:
                    arr2[ny][nx]=2
                    q.append((ny, nx))

    global result
    count = 0
    for i in range(n):
        for j in range(m):
            if arr2[i][j]==0:
                count += 1

    result = max(result, count)

result = 0
dfs(0)

print(result)

File: 백준/Gold/14502. 연구소/연구소.java
import java.util.*;
import java.io.*;

public class Main {
    static class Pair {
        int y, x;

        public Pair(int y, int x) {
            this.y = y;
            this.x = x;
        }
    }
    static int n, m;
    static int[][] arr;
    static boolean[][] visited;
    static int safeArea = Integer.MIN_VALUE;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        arr = new int[n][m];
        for (int i = 0; i<n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j<m; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        dfs(0);
        System.out.println(safeArea);
    }

    public static void dfs(int level) {
        if (level==3) {
            int[][] copyArr = new int[n][m];
            for (int i = 0; i < n; i++) {
                copyArr[i] = arr[i].clone();
            }
            bfs(copyArr);
            return;
        }
        for (int i = 0; i<n; i++) {
            for (int j = 0; j<m; j++) {
                if (arr[i][j]==0) {
                    arr[i][j] = 1;
                    dfs(level+1);
                    arr[i][j] = 0;
                }
            }
        }
    }

    public static void bfs(int[][] copyArr) {
        int[] dy = new int[]{0, 1, 0, -1};
        int[] dx = new int[]{1, 0, -1, 0};

        visited = new boolean[n][m];
        Queue<Pair> q = new LinkedList<>();

        for (int i = 0; i<n; i++) {
            for (int j = 0; j<m; j++) {
                if (copyArr[i][j]==2) {
                    q.add(new Pair(i, j));
                    visited[i][j] = true;
                }
            }
        }

        while (!q.isEmpty()) {
            Pair curr = q.poll();
            int y = curr.y, x = curr.x;

            for (int i = 0; i<4; i++) {
                int ny = y+dy[i];
                int nx = x+dx[i];
                if (ny>=0 && ny<n && nx>=0 && nx<m && copyArr[ny][nx]==0 && !visited[ny][nx]) {
                    visited[ny][nx] = true;
                    copyArr[ny][nx] = 2;
                    q.add(new Pair(ny, nx));
                }
            }
        }

        int count = 0;
        for (int i = 0; i<n; i++) {
            for (int j = 0; j<m; j++) {
                if (copyArr[i][j]==0 && !visited[i][j]) {
                    count++;
                }
            }
        }

        safeArea = Math.max(safeArea, count);
    }
}